runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,386 @@
+  return setmetatable({}, {
+    __index = global_state,
+    __newindex = function(state, key, value)
+      if global_state[key] == value then
+        rawset(state, key, nil)
+      else
+        rawset(state, key, value)
+      end
+    end,
+  })
+end)
+
+--- @inlinedoc
+--- @class vim.lsp.document_color.enable.Opts
+---
+--- Highlight style. It can be one of the pre-defined styles, a string to be used as virtual text, or a
+--- function that receives the buffer handle, the range (start line, start col, end line, end col) and
+--- the resolved hex color.
+--- Defaults to 'background'.
+--- @field style? string|fun(bufnr: integer, range: Range4, hex_code: string)
+
+-- Default options.
+--- @type vim.lsp.document_color.enable.Opts
+M._opts = { style = 'background' }
+
+--- @param color string
+local function get_contrast_color(color)
+  local r_s, g_s, b_s = color:match('^#(%x%x)(%x%x)(%x%x)$')
+  local r, g, b = tonumber(r_s, 16), tonumber(g_s, 16), tonumber(b_s, 16)
+
+  local luminance = 0.298912 * r + 0.586611 * g + 0.114478 * b
+  local is_bright_color = luminance > 127
+  return is_bright_color and '#000000' or '#ffffff'
+end
+
+--- Returns the hex string representing the given LSP color.
+--- @param color lsp.Color
+--- @return string
+local function get_hex_code(color)
+  -- The RGB values in lsp.Color are in the [0-1] range, but we want them to be in the [0-255] range instead.
+  --- @param n number
+  color = vim.tbl_map(function(n)
+    return math.floor((n * 255) + 0.5)
+  end, color)
+
+  return ('#%02x%02x%02x'):format(color.red, color.green, color.blue):lower()
+end
+
+--- Cache of the highlight groups that we've already created.
+--- @type table<string, true>
+local color_cache = {}
+
+--- Gets or creates the highlight group for the given LSP color information.
+---
+--- @param hex_code string
+--- @param style string
+--- @return string
+local function get_hl_group(hex_code, style)
+  if style ~= 'background' then
+    style = 'foreground'
+  end
+
+  local hl_name = ('LspDocumentColor_%s_%s'):format(hex_code:sub(2), style)
+
+  if not color_cache[hl_name] then
+    if style == 'background' then
+      api.nvim_set_hl(0, hl_name, { bg = hex_code, fg = get_contrast_color(hex_code) })
+    else
+      api.nvim_set_hl(0, hl_name, { fg = hex_code })
+    end
+
+    color_cache[hl_name] = true
+  end
+
+  return hl_name
+end
+
+--- @param bufnr integer
+local function buf_clear(bufnr)
+  local bufstate = bufstates[bufnr]
+  local client_ids = vim.tbl_keys((bufstate or {}).hl_infos) --- @type integer[]
+
+  for _, id in ipairs(client_ids) do
+    bufstate.hl_infos[id] = {}
+  end
+
+  api.nvim_buf_clear_namespace(bufnr, document_color_ns, 0, -1)
+  api.nvim__redraw({ buf = bufnr, valid = true, flush = false })
+end
+
+--- @param bufnr integer
+--- @param opts? vim.lsp.util._refresh.Opts
+local function buf_refresh(bufnr, opts)
+  opts = opts or {}
+  opts.bufnr = bufnr
+
+  util._refresh(ms.textDocument_documentColor, opts)
```

Comment:
I would explicitly pass the handler here, either by not using `_refresh` or adding a `handler` field to `opts`. This will allow you to avoid exposing the handler in this module and avoid adding the handler in `handlers.lua`.

Reply:
Good idea. I'll later refactor other modules to pass a local handler too.